File: src/api/client/repository/impl/fd/tg-msg.ts
import { MessageApi } from 'src/api/client/message-api';
import { FileDescAPIResponse } from 'src/api/client/model';
import { IFDRepository } from 'src/api/client/repository/interface';
import { MessageNotFound } from 'src/errors/telegram';
import { TGFSFileRef } from 'src/model/directory';
import { TGFSFile, TGFSFileVersion } from 'src/model/file';
import { Logger } from 'src/utils/logger';

export class TGMsgFDRepository implements IFDRepository {
  constructor(private readonly msgApi: MessageApi) {}

  public async save(
    fd: TGFSFile,
    messageId?: number,
  ): Promise<FileDescAPIResponse> {
    if (!messageId) {
      return {
        messageId: await this.msgApi.sendText(JSON.stringify(fd.toObject())),
        fd,
      };
    }
    // edit an existing message
    try {
      return await this.update(messageId, fd);
    } catch (err) {
      if (err instanceof MessageNotFound) {
        // the message to edit is gone
        return this.save(fd);
      } else {
        throw err;
      }
    }
  }

  private async update(
    messageId: number,
    fd: TGFSFile,
  ): Promise<FileDescAPIResponse> {
    return {
      messageId: await this.msgApi.editMessageText(
        messageId,
        JSON.stringify(fd.toObject()),
      ),
      fd,
    };
  }

  public async get(fileRef: TGFSFileRef): Promise<TGFSFile> {
    const message = (
      await this.msgApi.getMessages([fileRef.getMessageId()])
    )[0];

    if (!message) {
      Logger.error(
        `File description (messageId: ${fileRef.getMessageId()}) for ${
          fileRef.name
        } not found`,
      );
      return TGFSFile.empty(`[Content Not Found]${fileRef.name}`);
    }

    const fd = TGFSFile.fromObject(JSON.parse(message.text));

    const versions = Object.values(fd.versions);

    const nonEmptyVersions = versions.filter(
      (version) => version.messageId != TGFSFileVersion.EMPTY_FILE,
    ); // may contain empty versions

    const versionsWithoutSizeInfo = nonEmptyVersions.filter(
      (version) => version.size == TGFSFileVersion.INVALID_FILE_SIZE,
    );

    const fileMessages = await this.msgApi.getMessages(
      versionsWithoutSizeInfo.map((version) => version.messageId),
    );

    versionsWithoutSizeInfo.forEach((version, i) => {
      const fileMessage = fileMessages[i];
      if (fileMessage) {
        version.size = Number(fileMessage.document.size);
      } else {
        version.setInvalid();
      }
    });

    if (versionsWithoutSizeInfo.length > 0) {
      return (await this.update(fileRef.getMessageId(), fd)).fd;
    }

    return fd;
  }
}

File: tgfs/api/client/repository/impl/fd/tg_msg.py
import json
import logging
from typing import Optional

from tgfs.api.client.api.message import MessageApi
from tgfs.api.client.api.model import FileDescAPIResponse
from tgfs.api.client.repository.interface import IFDRepository
from tgfs.errors.telegram import MessageNotFound
from tgfs.model.directory import TGFSFileRef
from tgfs.model.file import TGFSFile

logger = logging.getLogger(__name__)


class TGMsgFDRepository(IFDRepository):
    def __init__(self, message_api: MessageApi):
        self.__message_api = message_api

    async def save(
        self, fd: TGFSFile, fr: Optional[TGFSFileRef] = None
    ) -> FileDescAPIResponse:
        # If file referer is None, create a new file descriptor message.
        if fr is None:
            return FileDescAPIResponse(
                message_id=await self.__message_api.send_text(fd.to_json()),
                fd=fd,
            )

        # If file referer is provided, try to update the existing file descriptor.
        # But if the message is not found (probably got deleted manually), a new file descriptor will be created.
        try:
            return FileDescAPIResponse(
                message_id=await self.__message_api.edit_message_text(
                    message_id=fr.message_id, message=fd.to_json()
                ),
                fd=fd,
            )
        except MessageNotFound:
            return await self.save(fd)

    async def get(self, fr: TGFSFileRef) -> TGFSFile:
        message = (await self.__message_api.get_messages([fr.message_id]))[0]

        empty = TGFSFile.empty(fr.name)

        if not message:
            logging.error(
                f"File descriptor (message_id: {fr.message_id}) for {fr.name} not found"
            )
            return empty

        fd = TGFSFile.from_dict(json.loads(message.text))

        versions = fd.get_versions(exclude_empty=True)
        file_messages = await self.__message_api.get_messages(
            [version.message_id for version in versions if version.message_id]
        )

        for i, version in enumerate(versions):
            if (file_message := file_messages[i]) and file_message.document:
                version.size = file_message.document.size
            else:
                logger.warning(
                    f"File message for version {version.id} of {fr.name} not found"
                )
                version.set_invalid()

        return fd if versions else empty
